build_tools/cmake/iree_filter_test.cmake

```diff
@@ -0,0 +1,72 @@
+# Copyright 2023 The IREE Authors
+#
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+include(CMakeParseArguments)
+
+function(iree_filter_test)
+  cmake_parse_arguments(
+    _FUNC
+    ""
+    "RESULT_VAR_ENABLED;TIMEOUT;DRIVER"
+    "LABELS"
+    ${ARGN}
+  )
+
+  set(_ENABLED TRUE)
+
+  # Exclude hostonly tests when cross-compiling.
+  if(CMAKE_CROSSCOMPILING AND "hostonly" IN_LIST _FUNC_LABELS)
+    set(_ENABLED FALSE)
+  endif()
+
+  # Sanitizer-based exclusions.
+  # CUDA is considered incompatible with all sanitizers.
+  if(IREE_ENABLE_ASAN)
+    if (("noasan" IN_LIST _FUNC_LABELS) OR (_FUNC_DRIVER STREQUAL "cuda"))
+      set(_ENABLED FALSE)
+    endif()
+  endif()
+  if(IREE_ENABLE_TSAN)
+    if (("notsan" IN_LIST _FUNC_LABELS) OR (_FUNC_DRIVER STREQUAL "cuda"))
+      set(_ENABLED FALSE)
+    endif()
+  endif()
+
+  # Architecture-based exclusions.
+  if ((IREE_ARCH MATCHES "^riscv_") AND ("noriscv" IN_LIST _FUNC_LABELS))
+    set(_ENABLED FALSE)
+  endif()
+
+  # Timeout-based exclusions.
+  #
+  # Tests timeout values are "short", "moderate", "long". The IREE-wide default
+  # is "short".
+  # On very slow configurations, we only run "short" tests (the IREE-wide
+  # default), excluding "moderate" and "long". On other somewhat-slow configs,
+  # we run "short" and "moderate" tests, excluding "long".
```

Comment:
@GMNGeoffrey may have opinions on this. I'm personally fine with landing and iterating as needed.